src/main/java/com/google/idea/perf/TracerController.kt

```diff
@@ -16,266 +16,75 @@
 
 package com.google.idea.perf
 
-import com.google.idea.perf.util.formatNsInUs
+import com.google.idea.perf.util.formatNsInMs
 import com.intellij.openapi.Disposable
 import com.intellij.openapi.application.ApplicationManager.getApplication
-import com.intellij.openapi.application.invokeAndWaitIfNeeded
 import com.intellij.openapi.application.invokeLater
 import com.intellij.openapi.diagnostic.Logger
 import com.intellij.openapi.progress.ProgressIndicator
 import com.intellij.openapi.progress.ProgressManager
 import com.intellij.openapi.progress.util.ProgressIndicatorBase
-import com.intellij.openapi.project.ProjectManager
-import com.intellij.openapi.rd.attachChild
 import com.intellij.openapi.util.Computable
-import com.intellij.psi.PsiElementFinder
 import com.intellij.util.concurrency.AppExecutorUtil
-import com.intellij.util.ui.UIUtil
-import java.awt.image.BufferedImage
-import java.io.File
-import java.io.IOException
-import java.lang.instrument.UnmodifiableClassException
-import java.lang.reflect.Method
 import java.util.concurrent.TimeUnit.MILLISECONDS
 import java.util.concurrent.atomic.AtomicBoolean
-import javax.imageio.ImageIO
-import kotlin.reflect.jvm.javaMethod
 
-// Things to improve:
-// - Audit overall overhead and memory usage.
-// - Make sure CPU overhead is minimal when the tracer window is not showing.
-// - Add some logging.
-// - Detect repeated instrumentation requests with the same spec.
-// - Add a proper progress indicator (subclass of ProgressIndicator) which displays text status.
-// - Consider moving callTree to CallTreeManager so that it persists after the Tracer window closes.
-// - Add visual indicator in UI if agent is not available.
-// - Make sure we're handling inner classes correctly (and lambdas, etc.)
-// - Try to reduce the overhead of transforming classes by batching or parallelizing.
-// - Support line-number-based tracepoints.
-// - Corner case: classes may being loaded *during* a request to instrument a method.
-// - Show all instrumented methods in the method list view, even if there are no calls yet(?)
-// - Cancelable progress indicator for class retransformations.
-
-class TracerController(
```

Comment:
At first glance it seems this base class might be overkill, because there's only a bit of code shared between the subclasses (mainly the data refresh loop). What do you think? I'm OK keeping it like this for now if preferred.

Reply:
Yeah, it could be overkill. My main reason for doing this was to reduce clutter and only show the code I want to see within derived classes. It does make code for me easier to read, but I'm also fine with removing the base class.